File: Progammers/연습문제/Level 1/신규 아이디 추천/Solution.java
class Solution {
    public String solution(String new_id) {
        String step_1 = new_id.toLowerCase();
        String step_2 = step_1.replaceAll("[^a-z0-9-_.]", "");
        String step_3 = step_2.replaceAll("[.]+", ".");
        String step_4 = step_3.replaceAll("^[.]|[.]$", "");
        String step_5 = step_4;
        if (step_4.equals("")){
            step_5 = "a";
        }
        String step_6 = step_5;
        if (step_6.length() >= 16) {
            step_6 = step_6.substring(0, 15);
            step_6 = step_6.replaceAll("^[.]|[.]$", "");
        }
        String step_7 = step_6;
        if (step_6.length() <= 2) {
            while (step_7.length() < 3) {
                step_7 += step_7.charAt(step_7.length() - 1);
            }
        }
        return step_7;
    }
}

File: Progammers/연습문제/Level 1/신규 아이디 추천/Solution.js
function solution(new_id) {
    let id = step1(new_id);;
    id = step2(id);
    id = step3(id);
    id = step4(id);
    id = step5(id);
    id = step6(id);
    id = step7(id);
    return id;
}

function step1(str) {
    return str.toLowerCase();
}

function step2(str) {
    const regex = new RegExp('[^a-z0-9-_.]', 'g');
    const result = str.replace(regex, '');
    return result;
}

function step3(str) {
    const regex = new RegExp('[.]{2,}', 'g');
    const result = str.replace(regex, '.');
    return result
}

function step4(str) {
    const regex = new RegExp('^\\.|\\.$');
    result = str.replace(regex, '');
    return result;
}

function step5(str) {
    return str || 'a';
}

function step6(str) {
    let result = str.substr(0, 15);

    const regex = new RegExp('\\.$');
    result = result.replace(regex, '');
    return result;
}

function step7(str) {
    str = str.padEnd(3, str[str.length-1]);
    return str;
}

// TC
new_id = "...!@BaT#*..y.abcdefghijklm";
console.log(solution(new_id));
